src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs

```diff
@@ -0,0 +1,232 @@
+        }
+
+        [JsonRpcMethod(Methods.TextDocumentImplementationName)]
+        public async Task<object> GetTextDocumentImplementationsAsync(JToken input, CancellationToken cancellationToken)
+        {
+            var textDocumentPositionParams = input.ToObject<TextDocumentPositionParams>();
+            return await this._protocol.FindImplementationsAsync(_workspace.CurrentSolution, textDocumentPositionParams, _clientCapabilities, cancellationToken).ConfigureAwait(false);
+        }
+
+        [JsonRpcMethod(Methods.TextDocumentRangeFormattingName)]
+        public async Task<TextEdit[]> GetTextDocumentRangeFormattingAsync(JToken input, CancellationToken cancellationToken)
+        {
+            var documentRangeFormattingParams = input.ToObject<DocumentRangeFormattingParams>();
+            return await this._protocol.FormatDocumentRangeAsync(_workspace.CurrentSolution, documentRangeFormattingParams, _clientCapabilities, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// FAR calls to get third party references (xaml) and to get those references it expectes to be called on the UI thread.
+        /// <see cref="AbstractFindUsagesService.FindReferencesAsync"/>
+        /// </summary>
+        [JsonRpcMethod(Methods.TextDocumentReferencesName)]
+        public async Task<object[]> GetTextDocumentReferencesAsync(JToken input, CancellationToken cancellationToken)
+        {
+            var referenceParams = input.ToObject<ReferenceParams>();
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+            return await this._protocol.FindReferencesAsync(_workspace.CurrentSolution, referenceParams, _clientCapabilities, cancellationToken).ConfigureAwait(false);
+        }
+
+        [JsonRpcMethod(Methods.TextDocumentSignatureHelpName)]
+        public async Task<SignatureHelp> GetTextDocumentSignatureHelpAsync(JToken input, CancellationToken cancellationToken)
+        {
+            var textDocumentPositionParams = input.ToObject<TextDocumentPositionParams>();
+            return await this._protocol.GetSignatureHelpAsync(_workspace.CurrentSolution, textDocumentPositionParams, _clientCapabilities, cancellationToken).ConfigureAwait(false);
+        }
+
+        [JsonRpcMethod(Methods.WorkspaceSymbolName)]
+        public async Task<SymbolInformation[]> GetWorkspaceSymbolsAsync(JToken input, CancellationToken cancellationToken)
+        {
+            var workspaceSymbolParams = input.ToObject<WorkspaceSymbolParams>();
+            return await this._protocol.GetWorkspaceSymbolsAsync(_workspace.CurrentSolution, workspaceSymbolParams, _clientCapabilities, cancellationToken).ConfigureAwait(false);
+        }
+
+#pragma warning disable VSTHRD100 // Avoid async void methods
+        private async void DiagnosticService_DiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs e)
+#pragma warning restore VSTHRD100 // Avoid async void methods
```

Comment:
this is bad.  an async void method means we also just lose all testability here.  i.e. we have async work happening, and we're not tracking it at all with our test helpers.  we need to address this.

Reply:
see NavigateToItemProvider.StartSearch for proper way to implement a void method that needs to do async work properly.

Reply:
Filed https://github.com/dotnet/roslyn/issues/40388 to properly cancel diagnostics.  This code was moved.  Already have a lot going on and a lot to test in this PR without also changing functionality, but will followup.